src/ripple/nodestore/impl/DatabaseShardImp.cpp

```diff
@@ -1419,6 +1627,273 @@ DatabaseShardImp::removeFailedShard(std::shared_ptr<Shard> shard)
+
+        auto const separateHistoricalPath = !historicalPaths_.empty();
+
+        //////////////////////////////////////////////////////////////////
+        // Lambdas used inside of this method:
+        auto const removeShard =
+            [this](std::uint32_t const shardIndex) -> void {
+            canAdd_ = false;
+
+            if (auto it = shards_.find(shardIndex); it != shards_.end())
+            {
+                if (it->second.shard)
+                    removeFailedShard(it->second.shard);
+                else
+                {
+                    JLOG(j_.warn()) << "can't find shard to remove";
+                }
+            }
+            else
+            {
+                JLOG(j_.warn()) << "can't find shard to remove";
+            }
+        };
+
+        auto const keepShard =
+            [this, &lock, removeShard, separateHistoricalPath](
+                std::uint32_t const shardIndex) -> bool {
+            if (countHistoricalShards(lock) >= maxHistoricalShards_)
+            {
+                JLOG(j_.error())
+                    << "maximum number of historical shards reached";
+
+                removeShard(shardIndex);
+                return false;
+            }
+            if (separateHistoricalPath &&
+                !sufficientStorage(1, PathDesignation::historical, lock))
+            {
+                JLOG(j_.error()) << "insufficient storage space available";
+
+                removeShard(shardIndex);
+                return false;
+            }
+
+            return true;
+        };
+
+        // Move a shard from the main shard path to a historical shard
+        // path by copying the contents, and creating a new shard.
+        auto const moveShard = [this,
+                                &lock](std::uint32_t const shardIndex) -> void {
+            auto const dst = chooseHistoricalPath(lock);
+
+            if (auto it = shards_.find(shardIndex); it != shards_.end())
+            {
+                if (auto& shard = it->second.shard; shard)
+                {
+                    // Close any open file descriptors before moving
+                    // the shard dir. Don't call removeOnDestroy since
+                    // that would attempt to close the fds after the
+                    // directory has been moved.
+                    shard->closeAll();
+
+                    try
+                    {
+                        // Move the shard directory to the new path
+                        boost::filesystem::rename(
+                            shard->getDir().string(),
+                            dst / std::to_string(shardIndex));
+                    }
+                    catch (std::exception const& e)
+                    {
+                        JLOG(j_.error())
+                            << "shard " << shardIndex
+                            << " failed to move to historical storage";
+
+                        return;
+                    }
```

Comment:
`e` is unused. Either remove it, or output `e.what()` as part of the error message.

Reply:
Done